### ISSUE TITLE
Method for selecting default executor

### ISSUE THREAD

**Issue by shahzebsiddiqui** [buildtesters/buildtest#251]
So here are few ideas on how we can select a default executor as per configuration.

Option 1: Add key ``default`` in executor class
--------------------------------------------------

It would be something like this

```
editor: vi
executors:
  default: runlocal
  normal:
    type: slurm
    description: submit jobs with slurm using srun instead of sbatch
    launcher: sbatch
    options:
      - "-p normal"
      - "-t 00:10"
      - "-N 1"
      - "-n 1"
  runlocal:
    type: local
    description: submit jobs locally
```

We would need to check value of ``default`` after the validate step against all executor objects since we won't know the executor names after the validate step. Json can't use ``enum`` in this case since the executor object name is not defined yet

Option 2: First executor object is default
----------------------------------------------
 
In this case, we don't need to add a separate key (``default``) to make the change. it's just a matter of order. In both case, the default executor does not have to be of ``type: local``. This is true, because we have not made strict check to ensure we have a local executor, if we wanted to we would have to change the json schema.

Option 3: Default per type
-----------------------------

Not sure if this would even make sense but having a default per executor type (``local``, ``slurm``). Here would be an example of two ``local`` and ``slurm`` executor definition with a default executor for each type

```
editor: vi
executors:
  default: 
    local: runlocal
    slurm: normal

  normal:
    type: slurm
    description: submit jobs with slurm using srun instead of sbatch
    launcher: sbatch
    options:
      - "-p normal"
      - "-t 00:10"
      - "-N 1"
      - "-n 1"
  admin:
    type: slurm
    description: run jobs on admin partition
    launcher: sbatch
    options:
      - "-p admin"
  runlocal:
    type: local
    description: submit jobs locally
  
  run_root:
    type: local
    description: run jobs that require root priviledges (root)
```

The behavior of buildtest can change as user change the default executor, it going to be important that their test are compatible when they make change to default executor.

**Comment by vsoch** [comment]
Since executors is a flat list, I like the idea of having a default, which (if not set) can default to the first entry (or some reasonable one determined by buildtest like local). There should also be a way to specify the executor of choice at any run time via a command line argument, `--executor` that takes the key and ignores the default setting.

**Comment by shahzebsiddiqui** [comment]
The ``--executor`` option can override the one in file so user can do ``buildtest build -c hpl.yml --executor admin`` to run HPL on ``admin`` executor. 

As per configuration would you go for Option 1 or Option 2. Either of them are doable for me. I don't see how Option 3 would work for time being.

**Comment by vsoch** [comment]
As I mentioned, option 1, but with an edit:

 - the default executor is provided by the "default" key under executors
 - if this key is not defined, we _either_ default to local or the first in the list.

I think I stated that above, sorry if it wasn't clear enough! I'll pick a number next time. :)

**Comment by shahzebsiddiqui** [comment]
Make sense to me, will go with option 1, and ``default`` can be an optional key and resort to first in list which is much simpler. If we pick a local executor, we can run into few issues that being

1. User didn't define a local executor, it's not enforced yet, so either we enforce it.
2. There can be more than one local executor so it won't be clear which executor to pick

**Comment by vsoch** [comment]
Just choose the first, spit out a warning.

**Comment by shahzebsiddiqui** [comment]
We decided to use ``Option 1`` and the default executor is now implemented in #252 so closing this issue for now. I will update the default configuration template shipped with buildtest to make use of ``default`` as the executor name.
